### ISSUE TITLE
WPF Painting Issue

### ISSUE THREAD

**Issue by jjxtra** [picoe/Eto#424]
The following code does not render anything to the screen. I may have a bug, but I've looked over the code many times and it looks OK...

```
        private Drawable d = new Drawable(false);
        private Bitmap layer = new Bitmap(2048, 1536, PixelFormat.Format32bppRgba);
        private Graphics layerGraphics;
        private Pen pen = new Pen(new Eto.Drawing.Color(0.0f, 0.0f, 0.0f, 1.0f), 10.0f);
        private PointF previousPoint;
        private bool mouseDown;

        public MainForm()
        {
            pen.LineJoin = PenLineJoin.Round;
            pen.LineCap = PenLineCap.Round;
            pen.MiterLimit = 4.0f;
            layerGraphics = new Graphics(layer);
            d.Paint += OnPaint;
            d.Invalidate();
            d.MouseDown += HandleMouseDown;
            d.MouseMove += HandleMouseMove;
            d.MouseUp += HandleMouseUp;
            Content = d;
        }

        private void HandleMouseDown(object sender, MouseEventArgs e)
        {
            if (!mouseDown)
            {
                mouseDown = true;
                previousPoint = new PointF(e.Location.X, e.Location.Y);
            }
        }

        private void HandleMouseMove(object sender, MouseEventArgs e)
        {
            if (!mouseDown)
            {
                return;
            }
            previousPoint = new PointF(e.Location.X, e.Location.Y);
            int x = Math.Min((int)previousPoint.X, (int)e.Location.X) - 10;
            int y = Math.Min((int)previousPoint.Y, (int)e.Location.Y) - 10;
            int xMax = Math.Max((int)previousPoint.X, (int)e.Location.X) + 10;
            int yMax = Math.Max((int)previousPoint.Y, (int)e.Location.Y) + 10;
            layerGraphics.DrawLine(pen, previousPoint.X, previousPoint.Y, e.Location.X, e.Location.Y);
            d.Invalidate(new Eto.Drawing.Rectangle(x, y, xMax - x, yMax - y));
        }

        private void HandleMouseUp(object sender, MouseEventArgs e)
        {
            mouseDown = false;
        }

        private void OnPaint(object sender, PaintEventArgs e)
        {
            Graphics g = e.Graphics;
            g.AntiAlias = true;
            g.DrawImage(layer, 0.0f, 0.0f);
        }
```

**Comment by jjxtra** [comment]
Calling Flush on the layerGraphics seems to have fixed it

**Comment by cwensley** [comment]
@jjxtra glad you found it!  I was just about to suggest that.  (;

**Comment by jjxtra** [comment]
Although now after a few seconds of drawing everything just goes blank and refuses to paint anything more. Events are still called normally. I switched to the Direct2D implementation and it doesn't seem to have the problem.

**Comment by cwensley** [comment]
Hm, ok I'll take a look.  For now you could change it so each drawing operation uses its own graphics object, as this is how it was initially designed to be used.  E.g.

```
using (var g = new Graphics(layer))
{
	g.DrawLine(...);
}
```

**Comment by jjxtra** [comment]
Switched to this, still has the bug...

```
            using (Graphics layerGraphics = new Graphics(layer))
            {
                layerGraphics.DrawLine(pen, previousPoint.X, previousPoint.Y, e.Location.X, e.Location.Y);
                layerGraphics.Flush();
            }
```

**Comment by cwensley** [comment]
Ah wait, I'm remembering a bug/memory leak in WPF that I haven't figured a way around yet.. Try adding `GC.Collect();` after the call to `Flush()`

**Comment by jjxtra** [comment]
That worked. Yucky. Maybe that hack should go in the WPF implementation until a better work around is found? Better yet, is the Direct2D implementation supported on Windows Store Apps? If so I'll just use that. I'm hoping to get You Doodle to Windows / MAC / Linux with this toolchain and would be happy to contribue bug reports, bug fixes, enhancements, etc.

**Comment by cwensley** [comment]
For reference, I ran into it when trying to use a double buffered Drawable [here](https://github.com/picoe/Eto/blob/develop/Source/Eto.Test/Eto.Test/Sections/Drawing/TextureBrushesSection2.cs#L65). 

It may be that I need to dispose the original bitmap [here](https://github.com/picoe/Eto/blob/develop/Source/Eto.Wpf/Drawing/GraphicsHandler.cs#L324) but haven't had time to look further into it yet.

**Comment by cwensley** [comment]
Yeah it is yucky.  I tried getting around this problem but it seemed that problem was due to objects allocated deep in the bowels of WPF that cause the issue.

I'm not sure about using Direct2D in windows store apps, though Eto uses SharpDX, so if that has been used in other windows store apps then it should be supported. 

With Eto, the Direct2D platform uses Windows Forms vs. WPF so there's that difference as well.

**Comment by jjxtra** [comment]
Should the GC.Collect go in the Dispose and Flush call for the GraphicsObject for the WPF implementation?

**Comment by cwensley** [comment]
Yeah that is probably the best solution to put it in the `GraphicsHandler.Close` method.  Just doing some quick google searches and it appears the only way around it is GC.Collect(). For example, [here](http://stackoverflow.com/questions/6713868/rendertargetbitmap-memory-leak)

**Comment by cwensley** [comment]
Ok this should now be fixed for the 2.2 release.  Thanks for reporting the issue!  

It's not ideal to call GC.Collect(), but it seems there's no other way at this point.
